src/main/scala/Hash.scala

```diff
@@ -1,19 +1,643 @@
 package chess
 
-import java.security.MessageDigest
+class Hash(
+    size: Int) {
 
-object Hash {
+  private def roleIndex(role: Role) = role match {
+    case Pawn   => 0
+    case Knight => 1
+    case Bishop => 2
+    case Rook   => 3
+    case Queen  => 4
+    case King   => 5
+  }
+
+  private def pieceIndex(piece: Piece) = piece.color match {
+    case Black => roleIndex(piece.role) * 2
+    case White => roleIndex(piece.role) * 2 + 1
+  }
+
+  private def posIndex(pos: Pos) =
+    8 * (pos.y - 1) + (pos.x - 1)
+
+  private def actorIndex(actor: Actor) =
+    64 * pieceIndex(actor.piece) + posIndex(actor.pos)
+
+  private def crazyPocketIndex(role: Role, color: Color, count: Int) = {
+    assert(0 < count && count <= 16)
+    assert(roleIndex(role) < 5)
```

Comment:
you won't find a single `assert` in scalachess or lila, because I don't like things that blow up at runtime.

Instead, you can express the uncompleteness of the function by returning an `Option[Int]`.